harness/determined/profiler.py

```diff
@@ -0,0 +1,517 @@
+
+LOG_NAMESPACE = "determined-profiler"
+
+
+
+
+class ProfilerAgent:
+    """
+    Agent that collects metrics and sends them to the master. It has:
+    - a thread to send data to the Master API (sender_thread)
+    - a thread to collect System Metrics and periodically flush to sender_thread (sys_metric_collector_thread)
+    - [UNIMPLEMENTED] something to batch Timings and periodically flush to sender_thread (timings_batcher)
+
+    The ProfilerAgent needs to be created at the beginning of training and it needs
+    to be notified every time the batch_idx increases. When it is created, it launches
+    the sender_thread and the sys_metric_collector_thread. They will be cleaned up
+    either once end_after_batch is finished or after 5 minutes have passed.
+
+    You can also ship Timings through the ProfilerAgent with the record_timing() method. This
+    functionality has not yet been implemented - we need to batch the timings and periodically
+    flush them to the sender_thread.
+
+    Profiling is only active between start_on_batch and end_after_batch. It will also automatically
+    shut down 5 minutes after starting. When profiling is not active, no system metrics are collected
+    and the record_timing function is a no-op.
+
+    Usage:
+    ```
+    profiler_agent = ProfilerAgent(self, trial_id, agent_id, master_url, start_on_batch, end_after_batch)
+
+    for batch_idx, batch in enumerate(batches):
+        profiler_agent.update_batch_idx(batch_idx)
+
+        # NOTE: Timing API has not been fully developed yet
+        forward_pass_timing = Timing("forward_pass")
+        forward_pass_timing.start()
+        # Do forward pass
+        forward_pass_timing.end()
+        profiler_agent.record_timing(forward_pass_timing)
+    ```
+    """
+    def __init__(self, trial_id: int, agent_id: str, master_url: str, start_on_batch: int, end_after_batch: Optional[int] = None):
+        self.current_batch_idx = 0
+        self.agent_id = agent_id
+        self.trial_id = trial_id
+        self.master_url = master_url
+        self.start_on_batch = start_on_batch
+        self.end_after_batch = end_after_batch
+
+        # Track duration to stop collecting after 5 minutes. start_time also serves as
+        # the indicator that collection has begun.
+        self.start_time = None
+        # TODO: Currently we only check for shutdown in update_batch_idx(). This is a problem when there
+        #       is a long period of time between batch_idx being updated (such as during validation)
+        #       because it won't shut down until validation completes. We probably need to create a
+        #       new thread to enforce the timeout correctly.
+        self.max_collection_seconds = 300
+        self.has_finished = False
+
+        # Set up the thread responsible for making API calls
+        self.send_queue = queue.Queue()
+        self.sender_thread = ProfilerSenderThread(self.send_queue, self.master_url)
+        self.sender_thread.start()
```

Comment:
Normally, python threds have a distinct "start/stop" lifetime which is a subset of their "init/gc" lifetime.  Or in other words, normally threads aren't started in `__init__()`, only created there.

Examples:
```python
# threading library
t = threading.Thread(target=my_func)
t.start()
t.join()
# gc happens later
```

Can we have a `ProfilingAgent.start()` method that starts both threads?  And a matching `ProfilingAgent.close()` method that (idempotently) stops/joins both threads?

Then the step after that is to write an `__enter__()` method that does `self.start(): return self` and an `__exit__()` that does `self.close` and you'll have a real nice API that will always shut down the threads it started, even in the case of uncaught exceptions.

Reply:
So if you don't mind, let's defer this problem. There is a PR that follows up on this one that hooks it into the TrialControllers. The current proposal is to use a context manager so we can move the start into that. If we decide not to use a context manager, then we can add a start() function

Reply:
Sure, I'm ok deferring this.  Although I think we should use a context manager and the context manager should call start() in it's `__enter__()` function.  That way the primary usage pattern is super easy, but you still give the flexibility to call it in whatever strange calling pattern you want (such as we needed to integrate it into a Keras Callback).

Reply:
Fixed